File: examples/nodejs/run_client.mjs
const URL = process.env.NODE_EXAMPLE_URL || "http://127.0.0.1:5200/orders";
const BODY = {item: "Buku Pemrograman", quantity: 1};


async function kirim(judul, headers) {
  const response = await fetch(URL, {
    method: "POST",
    headers: {
      "Authorization": "Bearer node-app-token",
      "Content-Type": "application/json",
      ...headers
    },
    body: JSON.stringify(BODY)
  });
  const data = await response.json();
  console.log(`\n${judul}`);
  console.log("HTTP       :", response.status);
  console.log("Keputusan  :", response.headers.get("x-ratf-decision"));
  console.log("Trust score:", data.ratf?.trust_score);
  console.log("Alasan     :", data.ratf?.reason_code);
  console.log("Respons    :", data);
}


await kirim("1. Request Node.js normal", {
  "X-Client-IP": "192.168.10.10",
  "X-Device-Id": "device-primary",
  "X-Hour-UTC": "10",
  "User-Agent": "MarketplaceNode/1.0"
});

await kirim("2. Token sama dari konteks yang sangat berbeda", {
  "X-Client-IP": "103.10.20.30",
  "X-Device-Id": "device-other",
  "X-Hour-UTC": "23",
  "User-Agent": "AutomationClient/4.0"
});

File: examples/flask_app/run_client.py
from datetime import datetime, timezone
import uuid

import requests


BASE_URL = "http://127.0.0.1:5100"
BODY = {"item": "Buku Pemrograman", "quantity": 1}


def tampilkan(judul, response):
    data = response.json()
    print(f"\n{judul}")
    print("HTTP       :", response.status_code)
    print("Keputusan  :", response.headers.get("X-RATF-Decision", data.get("decision")))
    print("Trust score:", response.headers.get("X-RATF-Score", data.get("trust_score")))
    print("Alasan     :", response.headers.get("X-RATF-Reason", data.get("reason_code")))
    print("Respons    :", data)


normal_headers = {
    "Authorization": "Bearer app-token-alice",
    "X-Client-Id": "marketplace-app",
    "X-Device-Id": "device-primary",
    "X-Request-Nonce": f"nonce_{uuid.uuid4().hex}",
    "Idempotency-Key": f"idem_{uuid.uuid4().hex}",
    "X-Experiment-Key": "local-experiment-key-32-characters-long",
    "X-Test-Source-IP": "192.168.10.10",
    "X-Test-Context-Time": datetime.now(timezone.utc).replace(hour=10).isoformat(),
    "User-Agent": "MarketplaceApp/1.0",
}

normal = requests.post(f"{BASE_URL}/api/orders", headers=normal_headers, json=BODY, timeout=5)
tampilkan("1. Request normal", normal)

changed_headers = normal_headers.copy()
changed_headers.update(
    {
        "X-Request-Nonce": f"nonce_{uuid.uuid4().hex}",
        "Idempotency-Key": f"idem_{uuid.uuid4().hex}",
        "X-Test-Source-IP": "103.10.20.30",
        "User-Agent": "AutomationClient/1.0",
    }
)
changed = requests.post(f"{BASE_URL}/api/orders", headers=changed_headers, json=BODY, timeout=5)
tampilkan("2. Token sama, tetapi IP dan aplikasi client berubah", changed)

replay = requests.post(f"{BASE_URL}/api/orders", headers=normal_headers, json=BODY, timeout=5)
tampilkan("3. Request normal dikirim ulang tanpa mengganti nonce", replay)

debug = requests.get(
    f"{BASE_URL}/app/debug/ratf",
    headers={"X-Debug-Key": "local-debug-key"},
    timeout=5,
).json()
print("\n4. State yang dapat dipakai pengembang untuk debug")
print("Storage              :", debug["storage_backend"])
print("Request allow tersimpan:", debug["context_history"].get("allowed_request_count", 0))
print("Context history      :", debug["context_history"])
print("Event terbaru        :", len(debug["recent_events"]))
print("Audit valid          :", debug["audit_integrity"]["valid"])
